File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
import re
def solution(str1, str2):
    list1 = []
    list2 = []
    for i in range(len(str1)-1):
        if re.match(r'^[A-Za-z]+$', str1[i]+str1[i+1]):
            list1.append(str1[i].upper()+str1[i+1].upper())
        
    for i in range(len(str2)-1):
        if re.match(r'^[A-Za-z]+$', str2[i]+str2[i+1]):
            list2.append(str2[i].upper()+str2[i+1].upper())

    
    alllist = 0
    onlylist = 0

    length = len(list1)
    
    for i in range(length-1, -1, -1):
        if list1[i] in list2:
            onlylist += 1
            list2.remove(list1[i])
            list1.pop()
            

    alllist = onlylist + len(list1) + len(list2)

    if alllist == 0:
        return 65536
    return int((onlylist/alllist)*65536)

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.*;
import java.util.regex.Pattern;

class Solution {
    public int solution(String str1, String str2) {
        
        ArrayList<String> string1 = new ArrayList();
        ArrayList<String> string2 = new ArrayList();
        
        str1 = str1.toUpperCase();
        str2 = str2.toUpperCase();
        
        for (int i=0; i<str1.length()-1; i++ ) {
            if (str1.substring(i, i+2).matches("[a-zA-Z]{2}")) {
                string1.add(str1.substring(i, i+2));
            }
        }
        
        for (int i=0; i<str2.length()-1; i++ ) {
            if (str2.substring(i, i+2).matches("[a-zA-Z]{2}")) {
                string2.add(str2.substring(i, i+2));
            }
        }
        
        if (string1.toArray().length == 0 && string2.toArray().length == 0) {
            return 1 * 65536;
        } else if (string1.toArray().length == 0 || string2.toArray().length == 0) {
            return 0;
        }
        
        int cnt = 0;
        
        ArrayList<String> temp = new ArrayList<>(string2);
        
        for (String i: string1) {
            if (temp.contains(i)) {
                cnt += 1;
                temp.remove(i);
            }
        }
        
        int cntAll = string1.size() + string2.size() - cnt;
        
        int answer = 65536 * cnt / cntAll;
        return answer;
    }
}